target/linux/realtek/files-6.6/drivers/i2c/busses/i2c-gpio-shared.c

```diff
@@ -0,0 +1,177 @@
+// SPDX-License-Identifier: GPL-2.0
+/*
+ * Bitbanging driver for multiple I2C busses with shared SCL pin using the GPIO API
+ * Copyright (c) 2025 Markus Stockhausen <markus.stockhausen at gmx.de>
+ */
+
+#include <linux/i2c-algo-bit.h>
+#include <linux/gpio/consumer.h>
+#include <linux/mod_devicetable.h>
+#include <linux/mutex.h>
+#include <linux/platform_device.h>
+
+#define SI2C_MAX_BUS	4
+
+struct si2c_ctx;
+
+struct si2c_bus {
+	int nr;
+	struct gpio_desc *sda;
+	struct i2c_adapter adap;
+	struct i2c_algo_bit_data bit_data;
+	struct si2c_ctx *ctx;
+};
+
+struct si2c_ctx {
+	struct device *dev;
+	struct gpio_desc *scl;
+	struct mutex lock;
+	struct si2c_bus bus[SI2C_MAX_BUS];
+};
+
+static void si2c_setsda(void *data, int state)
+{
+	struct si2c_bus *bus = data;
+
+	gpiod_set_value_cansleep(bus->sda, state);
+}
+
+static void si2c_setscl(void *data, int state)
+{
+	struct si2c_bus *bus = data;
+	struct si2c_ctx *ctx = bus->ctx;
+
+	gpiod_set_value_cansleep(ctx->scl, state);
+}
+
+static int si2c_getsda(void *data)
+{
+	struct si2c_bus *bus = data;
+
+	return gpiod_get_value_cansleep(bus->sda);
+}
+
+static int si2c_getscl(void *data)
+{
+	struct si2c_bus *bus = data;
+	struct si2c_ctx *ctx = bus->ctx;
+
+	return gpiod_get_value_cansleep(ctx->scl);
+}
+
+static int si2c_pre_xfer(struct i2c_adapter *adap)
+{
+	struct si2c_bus *bus = container_of(adap, typeof(*bus), adap);
+	struct si2c_ctx *ctx = bus->ctx;
+
+	dev_dbg(ctx->dev, "lock before transfer to bus %d\n", bus->nr);
+	mutex_lock(&ctx->lock);
+
+	return 0;
+}
+
+static void si2c_post_xfer(struct i2c_adapter *adap)
+{
+	struct si2c_bus *bus = container_of(adap, typeof(*bus), adap);
+	struct si2c_ctx *ctx = bus->ctx;
+
+	dev_dbg(ctx->dev, "unlock after transfer to bus %d\n", bus->nr);
+	mutex_unlock(&ctx->lock);
+}
+
+static int si2c_probe(struct platform_device *pdev)
+{
+	struct device *dev = &pdev->dev;
+	struct fwnode_handle *child;
+	struct si2c_ctx *ctx;
+	int tm, ret, nr = 0;
+
+	ctx = devm_kzalloc(dev, sizeof(*ctx), GFP_KERNEL);
+	if (!ctx)
+		return -ENOMEM;
+
+	ctx->dev = dev;
+	mutex_init(&ctx->lock);
+
+	ctx->scl =  devm_gpiod_get(dev, "scl", GPIOD_OUT_HIGH_OPEN_DRAIN);
+	if (IS_ERR(ctx->scl))
+		return dev_err_probe(dev, PTR_ERR(ctx->scl), "shared SCL node not found\n");
+
+	if (device_get_child_node_count(dev) >= SI2C_MAX_BUS)
+		return dev_err_probe(dev, -EINVAL, "Too many channels\n");
+
+	device_for_each_child_node(dev, child) {
+		struct si2c_bus *bus = &ctx->bus[nr];
+		struct i2c_adapter *adap = &bus->adap;
+		struct i2c_algo_bit_data *bit_data = &bus->bit_data;
+
+		bus->sda = devm_fwnode_gpiod_get(dev, child, "sda", GPIOD_OUT_HIGH_OPEN_DRAIN,
+						  fwnode_get_name(child));
+		if (IS_ERR(bus->sda))
+			return dev_err_probe(dev, PTR_ERR(ctx->scl),
```

Comment:
If you `return` here instead of `continue`ing, then all the other children will also not be initialized. I suggest to `fwnode_handle_put(child)`,log an error, and continue to the next sub-node.

Reply:
changed in new version